mlserver/codecs/decorator.py

```diff
@@ -0,0 +1,237 @@
+from functools import wraps, partial
+from typing import (
+    Any,
+    Callable,
+    Coroutine,
+    Dict,
+    List,
+    Optional,
+    Union,
+    Type,
+    Tuple,
+    get_type_hints,
+    TYPE_CHECKING,
+)
+
+try:
+    from typing import get_origin, get_args
+except ImportError:
+    # The `get_origin` and `get_args` don't exist in Python 3.7, so we'll
+    # backfill them manually.
+    # When we remove support for Python 3.7, remove these methods.
+    def get_origin(t: Type) -> Type:  # type: ignore
+        return getattr(t, "__origin__", type(None))
+
+    def get_args(t: Type) -> tuple:  # type: ignore
+        return getattr(t, "__args__", ())
+
+
+from ..types import InferenceRequest, InferenceResponse, ResponseOutput
+
+from .base import RequestCodec, InputCodec, find_input_codec, find_request_codec
+from .errors import InputsNotFound, OutputNotFound, CodecNotFound
+from .utils import Codec
+
+if TYPE_CHECKING:
+    from ..model import MLModel
+
+PredictFunc = Callable[
+    ["MLModel", InferenceRequest], Coroutine[Any, Any, InferenceResponse]
+]
+
+
+def _as_list(a: Optional[Union[Any, Tuple[Any]]]) -> List[Any]:
+    if a is None:
+        return []
+
+    if isinstance(a, tuple):
+        # Split into components
+        return list(a)
+
+    # Otherwise, assume it's a single element
+    return [a]
+
+
+def _is_codec_type(c: Codec, t: Type) -> bool:
+    if issubclass(c, t):  # type: ignore
+        return True
+
+    if isinstance(c, t):
+        return True
+
+    return False
+
+
+_is_input_codec = partial(_is_codec_type, t=InputCodec)
+_is_request_codec = partial(_is_codec_type, t=RequestCodec)
+
+
+def _is_optional(t: Type) -> bool:
+    origin = get_origin(t)
+    if origin == Optional:
+        return True
+
+    if origin == Union:
+        # Cover case where Optional[a] is reported as Union[a, None]
+        args = get_args(t)
+        if len(args) == 2 and type(None) in args:
+            return True
+
+    return False
+
+
+def _unwrap_optional(t: Type) -> Type:
+    args = get_args(t)
+    for arg in args:
+        if not isinstance(arg, type(None)):
+            return arg
+
+    return t
+
+
+class SignatureCodec(RequestCodec):
+    """
+    Internal codec that knows how to map type hints to codecs.
+    """
+
+    # TODO: Should this receive the whole class as argument?
```

Comment:
Initially I had the idea that, given that the `predict()` signature now gives all the info about what inputs a model takes (and which output it returns), we could also update the model's metadata to ensure it shows this info. That would probably need the whole `MLModel` class though, as it would need access to other bits and bobs from the model (and not just the `predict()` method). 

There's been already quite some time spent into this PR though, so I feel it's best to do that on a future iteration of this work.